design/src/smithy/event_streams.md

```diff
@@ -0,0 +1,302 @@
+The `Receiver` has more logic to handle buffering data and possibly errors.
+
+Server and client serialize similarly. Serializing for `CapturePokemonOperation` on the server, with `serialize_capture_pokemon_operation_response`:
+1. Sets the `content-type` header to `application/vnd.amazon.eventstream`
+2. Converts the `EventStreamSender` in the event stream structure into a `MessageStreamAdapter` with a marshaller for the error and data types
+   1. [This](https://github.com/awslabs/smithy-rs/blob/8f7e03ff8a84236955a65dba3d21c4bdbf17a9f4/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt#L511) is where it is generated
+3. Gives the body back to hyper
+```rust
+let body =
+   aws_smithy_http_server::body::boxed(aws_smithy_http_server::body::Body::wrap_stream({
+      let signer = aws_smithy_eventstream::frame::NoOpSigner {};
+      let error_marshaller =
+           crate::event_stream_serde::CapturePokemonOperationErrorMarshaller::new();
+       let marshaller = crate::event_stream_serde::CapturePokemonEventsMarshaller::new();
+       let adapter: aws_smithy_http::event_stream::MessageStreamAdapter<_, _> = output
+           .events
+           .into_body_stream(marshaller, error_marshaller, signer);
+       adapter
+   }));
+```
+The signer signs the message to be sent. [NoOpSigner](https://github.com/awslabs/smithy-rs/blob/8f7e03ff8a84236955a65dba3d21c4bdbf17a9f4/rust-runtime/aws-smithy-eventstream/src/frame.rs#L37) returns the message as is, `SigV4Signer` signs using the AWS SigV4 protocol. SigV4 requires an empty-payload signed message to be sent before effectively terminating the stream; to keep the same interface, `SignMessage::sign_empty` returns an `Option` to signal whether signing this last empty message is required.
+
+The marshallers set header and payload in a [Message](https://github.com/awslabs/smithy-rs/blob/8f7e03ff8a84236955a65dba3d21c4bdbf17a9f4/rust-runtime/aws-smithy-eventstream/src/frame.rs#L368) structure: `Message` is a structure with a vector, the headers; and bytes, the payload.
+The headers are the values targeted by the `@eventHeader` trait and the payload by `@eventPayload`.
+
+At the end of the marshalling and signing processes, `MessageStreamAdapter` takes the `Message` built by the marshaller and signer
+and [writes](https://github.com/awslabs/smithy-rs/blob/8f7e03ff8a84236955a65dba3d21c4bdbf17a9f4/rust-runtime/aws-smithy-eventstream/src/frame.rs#L224) it as bytes into a `Vec<u8>`,
+in a format of a sequence of bytes: `<type, data>` where `type` indicates if the `data` is a bool, integer and so on for all types.
+
+Headers that are always sent are:
+* `:message-type` (`event` or `exception`) to signal the kind of message being sent
+* `:content-type`, currently always set to `application/octet-stream`
```

Comment:
I don't think this is correct? If sending a `Blob` as the `@eventPayload`, then yes. But if it's a `structure`, then it will have the content type of the protocol (e.g., `application/json` for `@restJson1`).